core/arch/aarch64/codec.c

```diff
@@ -1739,6 +1766,89 @@ encode_opnd_prf12(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out
     return encode_opnd_mem12_scale(3, true, opnd, enc_out);
 }
 
+/* vindex_H: Index for vector with half elements (0-7). */
+
+static inline bool
+decode_opnd_vindex_H(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    uint bits = (enc >> 11 & 1) << 2 | (enc >> 21 & 1) << 1 | (enc >> 20 & 1) ;
+    *opnd = opnd_create_immed_int(bits, OPSZ_2b);
+    return true;
+}
+
+static inline bool
+encode_opnd_vindex_H(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
+{
+    ptr_int_t val;
+    if (!opnd_is_immed_int(opnd))
+        return false;
+    val = opnd_get_immed_int(opnd);
+    if (val < 0 || val >= 8)
+        return false;
+    *enc_out = (val >> 2 & 1) << 11 | (val >> 1 & 1) << 21 | (val & 1) << 20 ;
+    return true;
+}
+
+/* vindex_S: Index for vector with single elements (0-3). */
+
+static inline bool
+decode_opnd_vindex_S(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    uint bits;
+    ASSERT((enc >> 22 & 1) == 0 && "vindex_S should only be used for single width.");
```

Comment:
I don't understand why this isn't just a "return false". There is no other ASSERT in this set of functions. I wonder if there's a better way of handling this bit 22. Is it possible to combine vindex_S and vindex_D, with a '?' in bit 30 in the pattern in `codec.txt`?

Reply:
Right, I've updated it to use a single vindex_SD with '?'. I missed the fact that the codec.py script supports reading bits sets by other operands. Neat!